File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.cc
#include <iostream>
#include <iomanip>
#include <string>
#include <algorithm>
#include <vector>
#include <stack>
#include <queue>
#include <math.h>
#include <tuple>

#define MAX 1000000001
#define MIN -1000000001

using namespace std;
int n;
int arr[15];
int num[4];
int used[4];
int M = MIN, m = MAX;

int op(int a, int b, int i) {
	switch (i)
	{
	case 0:
		return a + b;
	case 1:
		return a - b;
	case 2:
		return a * b;
	case 3:
		return a / b;
	}
}

void func(int k, int sum) {
	if (k == n - 1) {
		M = max(M, sum);
		m = min(m, sum);
		return;
	}
	for (int i = 0; i < 4; i++)
	{
		if (num[i] > used[i]) {
			used[i]++;
			func(k + 1, op(sum, arr[k + 1], i));
			used[i]--;
		}
	}

}

int main(void) {
	ios::sync_with_stdio(0);
	cin.tie(0);

	cin >> n;
	for (int i = 0; i < n; i++)
	{
		cin >> arr[i];
	}
	for (int i = 0; i < 4; i++)
	{
		cin >> num[i];
	}
	func(0, arr[0]);
	cout << M << '\n' << m;
}

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
n = int(input())
arr = list(map(int, input().split()))
op = list(map(int, input().split()))
ops = [lambda x, y: x+y, lambda x,y : x-y, lambda x, y: x*y, lambda x, y : -(abs(x) // y) if x < 0  else x // y]
cnt = [0 for _ in range(4)]
import math
m = float("inf")
M = float("-inf")

def solve(idx, val):
    global op, cnt, m, M
    if(sum(op) == sum(cnt)):
        m = min(m, val)
        M = max(M, val)
        return    
    for i in range(4) :
        if cnt[i] < op[i]:
            cnt[i] += 1
            solve(idx+1, ops[i](val, arr[idx]))
            cnt[i] -= 1
        

solve(1, arr[0])        
print(M, m, sep='\n')
        
        
        
    